### DISCUSSION TITLE
Deliver slip / Invoices

### DISCUSSION THREAD

**Discussion by huwgit** [vendurehq/vendure#1980] [Q&A]
Trying to implement a delivery slip and invoice slip facility. Ideally a couple of buttons in the open order page. where do I start? Where can I edit the order page.  The delivery slip would display order details and shipping information.

**Comment by michaelbromley** [comment]
There is an existing plugin for invoices you can find here: https://pinelab-plugins.com/plugin/vendure-plugin-invoices

However, if you want to build you own, you can see how to add buttons to the order detail page here: https://www.vendure.io/docs/plugins/extending-the-admin-ui/add-actions-to-pages/

**Reply by huwgit** [reply]
I'm just learning this, definitely not one of my strengths:

I thought I'd try the example https://www.vendure.io/docs/plugins/extending-the-admin-ui/using-other-frameworks/ - thinking I might be able to create a delivery slip page. 

I got to the stage 5 - build the extension. I don't have yarn installed so Im using npm run build instead in the react-app directory. I'm getting the following errors:

**admin-ui/src/extension.routes.ts:3:32 - error TS2307: Cannot find module './extensions/895e207b259380711fb88f2a28ee65012d3f48b188243bf6f12198a259de5231/react-extension.module' or its corresponding type declarations.

3     loadChildren: () => import('./extensions/895e207b259380711fb88f2a28ee65012d3f48b188243bf6f12198a259de5231/react-extension.module').then(m => m.ReactUiExtensionModule),
                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~**

any ideas?

**Reply by michaelbromley** [reply]
It's hard to know exactly, but you can compare it to the example here and maybe it helps you figure out what is going wrong: https://github.com/vendure-ecommerce/real-world-vendure/tree/master/src/ui-extensions

**Reply by huwgit** [reply]
Taken from https://www.vendure.io/docs/plugins/extending-the-admin-ui/add-actions-to-pages/

Going back to the first post (adding an action button)  - would the following go in vendure-vonfig.ts: I don't have a folder called @vendure/admin-ui but I do have @vendure/admin-ui-plugin 

import { NgModule } from '@angular/core';
import { SharedModule, addActionBarItem } from '@vendure/admin-ui/core';

@NgModule({
  imports: [SharedModule],
  providers: [
    addActionBarItem({
      id: 'product-reviews',
      label: 'Product reviews',
      locationId: 'product-detail',
      buttonStyle: 'outline',
      routerLink: route => {
          const id = route.snapshot.params.id;
          return ['./extensions/reviews', id];
      },
      requiresPermission: 'SuperAdmin'
    }),
  ],
})
export class SharedExtensionModule {}

**Reply by michaelbromley** [reply]
The `@vendure/admin-ui` package is a dependency of `@vendure/ui-devkit` which you should have installed ([see docs](https://www.vendure.io/docs/plugins/extending-the-admin-ui/using-angular/#1-install-vendureui-devkit))

**Reply by huwgit** [reply]
I managed to install it on the server but failed on my laptop. On the server when I npm run dev I get at the end...

[server] info 1/19/23, 10:34 AM - [Vendure Server] Asset server: http://localhost:3000/assets 
[server] info 1/19/23, 10:34 AM - [Vendure Server] Dev mailbox:  http://localhost:3000/mailbox 
[server] info 1/19/23, 10:34 AM - [Vendure Server] Admin UI:     http://localhost:3000/admin 
[server] info 1/19/23, 10:34 AM - [Vendure Server] ================================================ 
[server] 
[server] > build
[server] > ng build --configuration production "--outputPath=/root/my-app/admin-ui/dist" "--base-href=/admin/"
[server] 
[server] Node.js version v17.9.0 detected.
[server] Odd numbered Node.js versions will not enter LTS status and should not be used for production. For more information, please see https://nodejs.org/en/about/releases/.
[server] Option "extractCss" is deprecated: Deprecated since version 11.0. No longer required to disable CSS extraction for HMR.


shop-api is working but on admin I get ...

{"statusCode":404,"message":"Cannot GET /admin-ui","timestamp":"2023-01-19T10:36:54.590Z","path":"/admin-ui"}

**Reply by michaelbromley** [reply]
Sometimes it takes up to a few minutes to compile the custom ui parts.

**Reply by huwgit** [reply]
What does this mean from above?  **Odd numbered Node.js versions will not enter LTS status ...**

**Reply by michaelbromley** [reply]
"Odd numbered Node.js versions will not enter LTS status"

This is covered here: https://nodejs.dev/en/about/releases/

In short - use even-numbered Node.js versions (16, 18 etc) rather than odd-numbered.
